lib/Dialect/FIRRTL/Transforms/CreateSiFiveMetadata.cpp

```diff
@@ -38,128 +39,281 @@ using circt::igraph::InstancePath;
+
+    // memFields[memFields.size() - 1] = hierPaths;
+    auto object = builderOM.create<ObjectOp>(memorySchemaClass, mem.getName());
+    for (auto field : llvm::enumerate(memoryParamNames)) {
+      auto propVal = createConstField(
+          llvm::StringSwitch<TypedAttr>(field.value())
+              .Case("name", builderOM.getStringAttr(mem.getName()))
               .Case("depth", mem.getDepthAttr())
               .Case("width", mem.getDataWidthAttr())
               .Case("maskBits", mem.getMaskBitsAttr())
               .Case("readPorts", mem.getNumReadPortsAttr())
               .Case("writePorts", mem.getNumWritePortsAttr())
               .Case("readwritePorts", mem.getNumReadWritePortsAttr())
               .Case("readLatency", mem.getReadLatencyAttr())
-              .Case("writeLatency", mem.getWriteLatencyAttr())));
+              .Case("writeLatency", mem.getWriteLatencyAttr())
+              .Case("hierarchy", {}));
+      if (!propVal)
+        propVal = hierpaths;
+
+      auto inPort =
+          builderOM.create<ObjectSubfieldOp>(object, 2 * field.index());
+      builderOM.create<PropAssignOp>(inPort, propVal);
+    }
+    auto portIndex = memoryMetadataClass.getNumPorts();
+    SmallVector<std::pair<unsigned, PortInfo>> newPorts = {
+        {portIndex, PortInfo(builderOM.getStringAttr(mem.getName() + "_field"),
+                             object.getType(), Direction::Out)}};
+    memoryMetadataClass.insertPorts(newPorts);
+    auto blockarg = memoryMetadataClass.getBodyBlock()->addArgument(
+        object.getType(), mem->getLoc());
+    builderOM.create<PropAssignOp>(blockarg, object);
+  }
 
-    auto object = builderOM.create<om::ObjectOp>(memorySchemaClass, memFields);
-    builderOM.create<om::ClassFieldOp>(
-        builderOM.getStringAttr("mem_" + mem.getName()), object);
+  void instantiateSifiveMetadata(FModuleOp topMod) {
+    if (!blackBoxMetadataClass && !memoryMetadataClass &&
+        !retimeModulesMetadataClass)
+      return;
+    auto builder = mlir::ImplicitLocOpBuilder::atBlockEnd(
+        mlir::UnknownLoc::get(circtOp->getContext()), circtOp.getBodyBlock());
+    SmallVector<PortInfo> mports;
+    auto sifiveMetadataClass = builder.create<ClassOp>(
+        builder.getStringAttr("SiFive_Metadata"), mports);
+    builder.setInsertionPointToStart(sifiveMetadataClass.getBodyBlock());
+
+    if (blackBoxMetadataClass)
+      builder.create<ObjectOp>(blackBoxMetadataClass,
+                               builder.getStringAttr("blackbox_metadata"));
+    if (memoryMetadataClass)
+      builder.create<ObjectOp>(memoryMetadataClass,
+                               builder.getStringAttr("memory_metadata"));
+
+    if (retimeModulesMetadataClass)
+      builder.create<ObjectOp>(
+          retimeModulesMetadataClass,
+          builder.getStringAttr("retime_modules_metadata"));
+
+    builder.setInsertionPointToEnd(topMod.getBodyBlock());
+    builder.create<ObjectOp>(sifiveMetadataClass,
```

Comment:
Should a reference to this object become an output of the topMod? In the current output, I don't think anything references the object.

Reply:
Do we want to update the top module signature to return the metadata object ? You are right, nothing references it now, we can define the ABI for the python script to interpret it correctly.

Reply:
If the ABI is "instantiate the SiFive_Metadata class", that would be reasonable to me. And in that case, we don't even need to instantiate the class within the top-module, we can leave the class as the "entrypoint".